File: linked-list/python/src/linked_list/linked_list.py
from __future__ import annotations

from typing import Generic, Optional, TypeVar

T = TypeVar("T")


class _Node(Generic[T]):
    __slots__ = ("value", "next")

    def __init__(self, value: T, next: Optional["_Node[T]"] = None) -> None:
        self.value = value
        self.next = next


class LinkedList(Generic[T]):
    def __init__(self) -> None:
        self._head: Optional[_Node[T]] = None
        self._count: int = 0

    def size(self) -> int:
        return self._count

    def append(self, value: T) -> None:
        node = _Node(value)
        if self._head is None:
            self._head = node
        else:
            current = self._head
            while current.next is not None:
                current = current.next
            current.next = node
        self._count += 1

    def prepend(self, value: T) -> None:
        self._head = _Node(value, self._head)
        self._count += 1

    def get(self, index: int) -> T:
        return self._node_at(index).value

    def remove(self, index: int) -> T:
        if index < 0 or index >= self._count:
            raise _out_of_range(index)
        if index == 0:
            assert self._head is not None
            removed = self._head
            self._head = removed.next
        else:
            previous = self._node_at(index - 1)
            assert previous.next is not None
            removed = previous.next
            previous.next = removed.next
        self._count -= 1
        return removed.value

    def insert_at(self, index: int, value: T) -> None:
        if index < 0 or index > self._count:
            raise _out_of_range(index)
        if index == 0:
            self.prepend(value)
            return
        if index == self._count:
            self.append(value)
            return
        previous = self._node_at(index - 1)
        previous.next = _Node(value, previous.next)
        self._count += 1

    def contains(self, value: T) -> bool:
        return self.index_of(value) >= 0

    def index_of(self, value: T) -> int:
        i = 0
        current = self._head
        while current is not None:
            if current.value == value:
                return i
            current = current.next
            i += 1
        return -1

    def to_array(self) -> list[T]:
        result: list[T] = []
        current = self._head
        while current is not None:
            result.append(current.value)
            current = current.next
        return result

    def _node_at(self, index: int) -> _Node[T]:
        if index < 0 or index >= self._count:
            raise _out_of_range(index)
        assert self._head is not None
        current = self._head
        for _ in range(index):
            assert current.next is not None
            current = current.next
        return current


def _out_of_range(index: int) -> IndexError:
    return IndexError(f"index out of range: {index}")

File: linked-list/typescript/src/linkedList.ts
class Node<T> {
  constructor(
    public value: T,
    public next: Node<T> | null = null,
  ) {}
}

export class LinkedList<T> {
  private head: Node<T> | null = null;
  private count = 0;

  size(): number {
    return this.count;
  }

  append(value: T): void {
    const node = new Node(value);
    if (this.head === null) {
      this.head = node;
    } else {
      let current = this.head;
      while (current.next !== null) current = current.next;
      current.next = node;
    }
    this.count++;
  }

  prepend(value: T): void {
    this.head = new Node(value, this.head);
    this.count++;
  }

  get(index: number): T {
    return this.nodeAt(index).value;
  }

  remove(index: number): T {
    if (index < 0 || index >= this.count) throw outOfRange(index);
    let removed: Node<T>;
    if (index === 0) {
      removed = this.head as Node<T>;
      this.head = removed.next;
    } else {
      const previous = this.nodeAt(index - 1);
      removed = previous.next as Node<T>;
      previous.next = removed.next;
    }
    this.count--;
    return removed.value;
  }

  insertAt(index: number, value: T): void {
    if (index < 0 || index > this.count) throw outOfRange(index);
    if (index === 0) {
      this.prepend(value);
      return;
    }
    if (index === this.count) {
      this.append(value);
      return;
    }
    const previous = this.nodeAt(index - 1);
    previous.next = new Node(value, previous.next);
    this.count++;
  }

  contains(value: T): boolean {
    return this.indexOf(value) >= 0;
  }

  indexOf(value: T): number {
    let i = 0;
    for (let current = this.head; current !== null; current = current.next, i++) {
      if (current.value === value) return i;
    }
    return -1;
  }

  toArray(): T[] {
    const result: T[] = [];
    for (let current = this.head; current !== null; current = current.next) {
      result.push(current.value);
    }
    return result;
  }

  private nodeAt(index: number): Node<T> {
    if (index < 0 || index >= this.count) throw outOfRange(index);
    let current = this.head as Node<T>;
    for (let i = 0; i < index; i++) current = current.next as Node<T>;
    return current;
  }
}

function outOfRange(index: number): RangeError {
  return new RangeError(`index out of range: ${index}`);
}
